brave/src/main/java/brave/internal/baggage/UnsafeArrayMap.java

```diff
@@ -0,0 +1,378 @@
+      Map.Entry that = (Map.Entry) o;
+      for (int i = 0; i < toIndex; i += 2) {
+        if (that.getKey().equals(array[i]) && equal(that.getValue(), array[i + 1])) return true;
+      }
+      return false;
+    }
+  }
+
+  abstract class CollectionView<E> implements Collection<E> {
+    final int fromIndex, toIndex, size;
+
+    protected CollectionView(int fromIndex, int toIndex) {
+      this.fromIndex = fromIndex;
+      this.toIndex = toIndex;
+      this.size = calculateSize(fromIndex, toIndex);
+    }
+
+    int advancePastFiltered(int i) {
+      while (i < toIndex && isFilteredKey(i)) i += 2;
+      return i;
+    }
+
+    @Override public int size() {
+      return size;
+    }
+
+    abstract E elementAtArrayIndex(int i);
+
+    @Override public Iterator<E> iterator() {
+      return new ReadOnlyIterator();
+    }
+
+    @Override public Object[] toArray() {
+      return copyTo(new Object[size]);
+    }
+
+    @Override public <T> T[] toArray(T[] a) {
+      T[] result = a.length >= size ? a
+          : (T[]) Array.newInstance(a.getClass().getComponentType(), size());
+      return copyTo(result);
+    }
+
+    <T> T[] copyTo(T[] dest) {
+      for (int i = fromIndex, d = 0; i < toIndex; i += 2) {
+        if (isFilteredKey(i)) continue;
+        dest[d++] = (T) elementAtArrayIndex(i);
+      }
+      return dest;
+    }
+
+    final class ReadOnlyIterator implements Iterator<E> {
+      int i = advancePastFiltered(fromIndex);
+
+      @Override public boolean hasNext() {
+        i = advancePastFiltered(i);
+        return i < toIndex;
+      }
+
+      @Override public E next() {
+        if (!hasNext()) throw new NoSuchElementException();
+        E result = elementAtArrayIndex(i);
+        i += 2;
+        return result;
+      }
+
+      @Override public void remove() {
+        throw new UnsupportedOperationException();
+      }
+    }
+
+    @Override public boolean containsAll(Collection<?> c) {
+      if (c == null) return false;
+      if (c.isEmpty()) return true;
+
+      for (Object element : c) {
+        if (!contains(element)) return false;
+      }
+      return true;
+    }
+
+    @Override public boolean isEmpty() {
+      return size == 0;
+    }
+
+    @Override public boolean add(E e) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean remove(Object o) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean addAll(Collection<? extends E> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean retainAll(Collection<?> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean removeAll(Collection<?> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public void clear() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  static final class Entry<K, V> implements Map.Entry<K, V> {
```

Comment:
Didn't think about the performance implications, but what could we save and where would we lose some performance if implementing `AbstractMap`?

Reply:
yeah I think you answered your own question. we'd need to override their crappy impls like 

```
    public int size() {
        return entrySet().size();
    }
```

anyway next change that clears all the last relics of list is much smaller, so it might be moot

Reply:
the longer answer I can put in RATIONALE also. Basically all operations in abstract map are criminally inefficient if we think we are providing a cheap view over an array. everything allocates something. The neat thing here is most things are lazy. ex "containsX" methods don't have to allocate anything. we don't do things like allocate both an entryset and an iterator to implement get() etc. Finally, not extending abstractmap allows us to better verify that the code is doing what it should be doing (especially now that there is less code thanks to you!)

Reply:
added rationale!